src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -0,0 +1,980 @@
+    {
+        bool emit = false;
+        auto & tz = mTimeZoneObj.timeZoneList[0].timeZone;
+        if (tz.offset != lastTz.offset)
+        {
+            emit = true;
+        }
+        if ((tz.name.HasValue() && lastTz.name.HasValue()) && !(tz.name.Value().data_equal(lastTz.name.Value())))
+        {
+            emit = true;
+        }
+        if (emit)
+            mEventFlag = TimeSyncEventFlag::kTimeZoneStatus;
+    }
+    return mTimeSyncDataProvider.StoreTimeZone(GetTimeZone());
+}
+
+CHIP_ERROR TimeSynchronizationServer::LoadTimeZone()
+{
+    mTimeZoneObj.timeZoneList = Span<TimeZoneStore>(mTz);
+    for (auto & tzStore : mTimeZoneObj.timeZoneList)
+    {
+        memset(tzStore.name, 0, sizeof(tzStore.name));
+        tzStore.timeZone = { .offset = 0, .validAt = 0, .name = MakeOptional(CharSpan(tzStore.name, sizeof(tzStore.name))) };
+    }
+    return mTimeSyncDataProvider.LoadTimeZone(mTimeZoneObj);
+}
+
+CHIP_ERROR TimeSynchronizationServer::SetDSTOffset(const DataModel::DecodableList<Structs::DSTOffsetStruct::Type> & dstL)
+{
+    size_t items;
+    VerifyOrReturnError(CHIP_NO_ERROR == dstL.ComputeSize(&items), CHIP_IM_GLOBAL_STATUS(InvalidCommand));
+
+    if (items > CHIP_CONFIG_DST_OFFSET_LIST_MAX_SIZE)
+    {
+        return CHIP_ERROR_BUFFER_TOO_SMALL;
+    }
+
+    if (items == 0)
+    {
+        return ClearDSTOffset();
+    }
+
+    auto newDstL   = dstL.begin();
+    auto & dstList = mDstOffsetObj.dstOffsetList;
+    size_t i       = 0;
+
+    while (newDstL.Next())
+    {
+        dstList[i] = newDstL.GetValue();
+        i++;
+    }
+
+    if (CHIP_NO_ERROR != newDstL.GetStatus())
+    {
+        mTimeSyncDataProvider.LoadDSTOffset(mDstOffsetObj);
+        return CHIP_IM_GLOBAL_STATUS(InvalidCommand);
+    }
+
+    mDstOffsetObj.size = i;
+
+    // only 1 validuntil null value and shall be last in the list
+    uint64_t lastValidUntil = 0;
+    for (i = 0; i < mDstOffsetObj.size; i++)
+    {
+        const auto & dstItem = GetDSTOffset()[i];
+        // list should be sorted by validStarting
+        // validUntil shall be larger than validStarting
+        if (!dstItem.validUntil.IsNull() && dstItem.validStarting >= dstItem.validUntil.Value())
+        {
+            mTimeSyncDataProvider.LoadDSTOffset(mDstOffsetObj);
+            return CHIP_ERROR_INVALID_TIME;
+        }
+        // validStarting shall not be smaller than validUntil of previous entry
+        if (dstItem.validStarting < lastValidUntil)
+        {
+            mTimeSyncDataProvider.LoadDSTOffset(mDstOffsetObj);
+            return CHIP_ERROR_INVALID_TIME;
+        }
+        lastValidUntil = !dstItem.validUntil.IsNull() ? dstItem.validUntil.Value() : lastValidUntil;
+        // only 1 validUntil null value and shall be last in the list
+        if (dstItem.validUntil.IsNull() && (i != mDstOffsetObj.size - 1))
+        {
+            mTimeSyncDataProvider.LoadDSTOffset(mDstOffsetObj);
```

Comment:
So I wonder: would it make sense to move these `LoadDSTOffset` calls into the caller of SetDSTOffset, so that we don't have to keep repeating them?  Similar for the LoadTimeZone bits in SetTimeZone...  Might save some flash.

Followup is very much fine for this.

Reply:
Tracked in https://github.com/project-chip/connectedhomeip/issues/27575, please follow up in issue, resolving here.